src/main/java/duke/TaskList.java

```diff
@@ -0,0 +1,93 @@
+package duke;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collectors;
+
+/**
+ * Utility wrapper class that makes working with task lists easier.
+ */
+public class TaskList {
+    private List<Task> tasks;
+    private Storage storage;
+
+    TaskList(Storage storage, Ui ui) {
+        this.storage = storage;
+        try {
+            tasks = storage.loadTasks();
+        } catch (Exception e) {
+            ui.showError(e.getMessage());
+            tasks = new ArrayList<>();
+        }
+    }
+
+    /**
+     * Marks the task located at the given index.
+     *
+     * @param index Index of the task in the list.
+     * @throws DukeException If task list could not be saved.
+     */
+    public void mark(int index) throws DukeException {
+        Task task = tasks.get(index);
+        task.mark();
+        storage.saveTasks(tasks);
+    }
+
+    /**
+     * Unmarks the task located at the given index.
+     *
+     * @param index Index of the task in the list.
+     * @throws DukeException If task list could not be saved.
+     */
+    public void unmark(int index) throws DukeException {
+        Task task = tasks.get(index);
+        task.unmark();
+        storage.saveTasks(tasks);
+    }
+
+    /**
+     * Adds the given task to the list.
+     *
+     * @param task Task to be added.
+     * @throws DukeException If task list could not be saved.
+     */
+    public void add(Task task) throws DukeException {
+        tasks.add(task);
+        storage.saveTasks(tasks);
+    }
+
+    /**
+     * Removes the task located at the given index.
+     *
+     * @param index Index of the task in the list.
+     * @throws DukeException If task list could not be saved.
+     */
+    public void remove(int index) throws DukeException {
+        tasks.remove(index);
+        storage.saveTasks(tasks);
+    }
+
+    public List<Task> search(String keyword) {
+        return tasks.stream().filter(x -> x.getDescription().contains(keyword)).collect(Collectors.toList());
```

Comment:
Efficient solution for the search function, using streams is a clever implementation!